cloudinit/sources/DataSourceWSL.py

```diff
@@ -143,22 +143,35 @@ def candidate_user_data_file_names(instance_name) -> List[str]:
     ]
 
 
-def load_yaml_or_bin(data_path: str) -> Optional[Union[dict, bytes]]:
-    """
-    Tries to load a YAML file as a dict, otherwise returns the file's raw
-    binary contents as `bytes`. Returns `None` if no file is found.
-    """
-    try:
-        bin_data = util.load_binary_file(data_path)
-        dict_data = util.load_yaml(bin_data)
-        if dict_data is None:
-            return bin_data
+class ConfigData:
```

Comment:
It looks like with the introduction of the ConfigData class, and your use below, you are ultimately trying to determine, is the type of data cloud-config or not, and if not, pass it on to cloud-init for normal user-data processing as #include user-data. While adding ConfigDict seems to simplify typing expectations from load_ubuntu_pro_data , using this class definition seems add complexity to some of our early exit logic in `_get_data` that may invalidate our checks such as `if not any([user_data, agent_data]): return False`. I'm trying to think of a constructive way to approach to proposed instead of just nay-saying here, so I'd like to look through this tomorrow to provide a constructive patch.

Reply:
@CarlosNihelton: if `.ubuntupro/.cloud-init/<INSTANCE_NAME>.user-data` is an empty file or empty dict, and `.cloud-init/<INSTANCE_NAME>.user-data` exists. What is the desired behavior for cloud-init here? I'd vote that cloud-init prefer the locally-defined non-empty config in `.cloud-init/<INSTANCE_NAME>.user-data`. But, the code we have currently in this PR will not attempt to read the locally-defined  `.cloud-init/<INSTANCE_NAME>.user-data` because user-data will be a [ConfigData instance instead of `None` with a ConfigData._data == dict()](https://github.com/canonical/cloud-init/pull/5538/files#diff-ca288809c5796ee57b583399b253f98d43607212de7ab1f3d285ac2fa4828e82R323). In this case (empty .ubuntupro/...user-data file) we will ignore any applicable user-data.

Reply:
> What is the desired behavior for cloud-init here?

The empty .ubuntu-pro file indicates that the sysadmin wants no user-data ingested in the instance being created in response to a Landscape command to Ubuntu Pro for WSL. If we opt for ingesting whatever the local user may have defined, that could cause surprise to the sysadmin. When a user creates their own instances there would be no Landscape provided data matching up that instance by name. If it does exist, it means the organization is in control of such instance and thus users should not override the organization desired config for that instance.

If we find Landscape provided data for the instance being initialized, we should not even bother looking for user data. The organization must be in control in that case.

Reply:
That makes sense. And adding an additional debug log to tell us why we are ignoring .cloud-init/user-data will help inform future us when debugging a problem in the future.